frame/elections-phragmen/src/lib.rs

```diff
@@ -656,93 +751,135 @@ decl_module! {
 	///
-	/// If replacement exists, this will read and write from/into both `Members` and `RunnersUp`.
-	fn remove_and_replace_member(who: &T::AccountId) -> Result<bool, DispatchError> {
-		let mut members_with_stake = Self::members();
-		if let Ok(index) = members_with_stake.binary_search_by(|(ref m, ref _s)| m.cmp(who)) {
-			members_with_stake.remove(index);
-
-			let next_up = <RunnersUp<T>>::mutate(|runners_up| runners_up.pop());
-			let maybe_replacement = next_up.and_then(|(replacement, stake)|
-				members_with_stake.binary_search_by(|(ref m, ref _s)| m.cmp(&replacement))
+	/// ### Note
+	///
+	/// This function attempts to preserve the prime. If the removed members is not the prime, it is
+	/// set again via [`Trait::ChangeMembers` ] to prevent it being wiped.
+	fn remove_and_replace_member(who: &T::AccountId, slash: bool) -> Result<bool, DispatchError> {
+		// closure will return:
+		// - `Ok(Option(replacement))` if member was removed and replacement was replaced.
+		// - `Ok(None)` if member was removed but no replacement was found
+		// - `Err(_)` if who is not a member.
+		<Members<T>>::try_mutate(|members| {
+			// we remove the member anyhow, regardless of having a runner-up or not.
+			// TODO: a way to make sure all tests operate in a way that accounts are not sorted.
+			let remove_index = members
+				// note: members are always sorted.
+				.binary_search_by(|m| m.who.cmp(who))
+				.map_err(|_| Error::<T>::NotMember)?;
+			let removed = members.remove(remove_index);
+			if slash {
+				let (imbalance, _remaining) = T::Currency::slash_reserved(who, removed.deposit);
+				debug_assert_eq!(_remaining, 0u32.into());
+				T::LoserCandidate::on_unbalanced(imbalance);
+				Self::deposit_event(RawEvent::SeatHolderSlashed(who.clone(), removed.deposit));
+			} else {
+				T::Currency::unreserve(who, removed.deposit);
+			}
+
+			Ok(<RunnersUp<T>>::mutate(|r| r.pop()).map(|next_best| {
+				// invariant: Members and runners-up are disjoint. This will always be err and give
+				// us an index to insert.
+				members
+					.binary_search_by(|m| m.who.cmp(&next_best.who))
 					.err()
 					.map(|index| {
-						members_with_stake.insert(index, (replacement.clone(), stake));
-						replacement
-					})
-			);
-
-			<Members<T>>::put(&members_with_stake);
-			let members = members_with_stake.into_iter().map(|m| m.0).collect::<Vec<_>>();
-			let result = Ok(maybe_replacement.is_some());
-			let old = [who.clone()];
-			match maybe_replacement {
-				Some(new) => T::ChangeMembers::change_members_sorted(&[new], &old, &members),
-				None => T::ChangeMembers::change_members_sorted(&[], &old, &members),
+						members.insert(index, next_best.clone());
+					});
+				next_best
+			}))
+		})
+		.map(|maybe_replacement| {
+			let member_ids = Self::members()
+				.into_iter()
+				.map(|x| x.who.clone())
+				.collect::<Vec<_>>();
+			let outgoing = &[who.clone()];
+			let current_prime = T::ChangeMembers::get_prime().unwrap_or_default();
```

Comment:
this is wrong we don't want this function to set prime to default when there is no prime in the first place.

Reply:
Great point